### PR TITLE
[ISTHMUS] Support approx_count_distinct in substrait to sql conversion 

### PR DESCRIPTION
**Pull request by jinfengni**

### CODE REVIEW

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:11]: Minor nit - remove wildcard import. You may want to increase default in IntelliJ which annoyingly defaults to this after three imports from the same package.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:172]: Prepend "Unsupported " here so it says "Unsupported aggregation function ..."

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:206]: Consider factoring this check out to a helper method as it's repeated above.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:233]: What does "%" do as opposed to "%s" in format string? Just curious

File: isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java:126]: Remove System.out.println

File: isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java:125]: Stronger assert would be good her. Couple of things here would be nice:
- Compare equality of original SQL where you replace `count(distinct <col>)` with `approx_count_distinct(<col>)`
- Convert newSql back to Plan using buildPlanFromQuery and then to proto, and compare to newPlan converted to proto.

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:149]: If never using, remove this condition and argument. Best if we don't need it (though we might eventually)

**Comment by JamesRTaylor** [comment]
A few minor comments, but this looks very good to me.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:11]: done

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:172]: done.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:233]: Thanks for catching this bug.  My eyes were not sharp enough to recognize these typos.  Changed to `%s` in all error messages.

File: isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java:126]: done

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:206]: Factor the common codes into a private method.

File: isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/RelCopyOnWriteVisitorTest.java:125]: For the first equality checking of SQL:
   - I feel comparing text is not a very stable way, since any format changing in Calcite's  RelToSql class could lead to testcase failure. 
   - Also, the converted `newSQL` contains a different field name `$f0` for the expression. That issue will be addressed later on.  Before that is fixed, could not compare SQL directly. 
 
For the second checking of substrait pojo/proto plan : 
- This pr adds support of `approx_count_distinct()` in substrait rel -> sql.  But the other direction (`sql` -> `substrait rel`) does not have support of `approx_count_distinct()`.  That probably need a separate PR, as I feel it's not straightforward to fix it (requires changing the function resolution logic). 
- So, for now, I could not add the checking of pojo/proto plan.  I put TODO to add new checking once 'sql' -> substrait work properly. 
-

**Comment by jinfengni** [comment]
@JamesRTaylor thanks a lot for your review! I modified the pr based on your feedback.  Can you please take a look? Thanks!

**Comment by jacques-n** [comment]
This PR appears to have multiple changes in it, not just what is in the title. Can you fix?

**Comment by jinfengni** [comment]
I changed pr's title to reflect the changes made in this pr. 

Also, after changing to use Calcite's APPROX_COUNT_DISTINCT, and change the code of converting substrait function to Calcite SqlOperator,  we could convert back to `APPROX_COUNT_DISTINCT` directly; no need for special processing.  So, clean up some of earlier implementation.

**Comment by JamesRTaylor** [comment]
LGTM

**Comment by JamesRTaylor** [comment]
Would be good to rebase after merge of dependent PR and force push to get rid of merge commit.

**Comment by jinfengni** [comment]
I have a fix to address `approx_count_distinct` conversion in sql -> substrait.  The simple idea is to replace Calcite's `count` + distinct/approximate with `approx_count_distinct` before converting into substrait function. I'll rebase and clean up the code, and submit for review.   

@jacques-n , please do not merge this PR for now.  Thanks!

File: isthmus/src/test/java/io/substrait/isthmus/PlanTestBase.java
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/PlanTestBase.java:88]: Is this conversion just to see if you get an exception or not? Would it be possible to go back from sql to proto rel or pojo rel and compare it's the same?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitToSql.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitToSql.java:69]: Just curious on what the implications of these defaults are. Does this determine whether column name references are double quoted? What about the case sensitive case? A detail, but eventually we'll need to deal with that. Also, with SQL dialect set to SNOWFLAKE, does `supportsApproxCountDistinct()` still return false? That seems like a bug, no? Maybe some comments would be good here.

File: isthmus/src/main/java/io/substrait/isthmus/expression/AggregateFunctionConverter.java
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/expression/AggregateFunctionConverter.java:85]: Does this mean we can go from sql -> plan for approx_count_distinct?

**Comment by JamesRTaylor** [comment]
This looks really good. Just a couple of questions.

File: isthmus/src/test/java/io/substrait/isthmus/PlanTestBase.java
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/PlanTestBase.java:88]: Previously, this conversion is to return a sql and assert Sql contains certain substring. Later, I changed to return RelNode, and verify RelNode tree contains certain expression.  

I tried to go back from sql to pojo rel and do the compare. The problem is `toSql` may loose some information. Consider the following query:
```
select p_size  from part where p_partkey > cast(100 as bigint)
``` 

In the  first round sql -> RelNode -> substrait rel ->  Calcite rel,  the condition is essentially represented as
`p_partkey > 100:bigint`.  However, from calcite rel -> sql, the condition becomes `p_partkey > 100`.  If we convert such sql back to substrait rel, the expression would contains a `cast`. Although they should return a same resultset, the comparison of pojo rel would fail.

In another words,  `toSql()` produces semantic equivalent yet different query.  Probably,  I should remove this `toSql` step.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitToSql.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitToSql.java:69]: `.withIdentifierQuoteString` will determine how to quote an identifier. Passing a `null` means no quoted identifier. Passing `"` would double quote all identifiers, including table/column names. 

Calcite has a `SnowflakeDialect`[1]. It does set `supportsApproxCountDistinct` to true. However, it specifies double quote for all identifier, which would lead problem to table/column searching in catalog. 

I changed the code to just use Calcite's `EMPTY_CONTEXT` to construct a sql dilect, since `EMPTY_CONTEXT` defaults to use null as IdentifierQuoteString. Same as SnowflakeDialect, I set `supportsApproxCountDistinct` for the constructed sql dialect. 

1. https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java#L28-L29

File: isthmus/src/main/java/io/substrait/isthmus/expression/AggregateFunctionConverter.java
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/expression/AggregateFunctionConverter.java:85]: Yes, now we can go sql -> substrait for approx_count_distinct.  Previously, in sql -> substrait, the code will convert `approx_count_distinct` to `count`, which is not correct.

**Comment by JamesRTaylor** [comment]
LGTM

**Review by JamesRTaylor — APPROVED** [review]

**Review by jacques-n — APPROVED** [review]
LGTM.

**Pull request merged**
